### DISCUSSION TITLE
Don't drop source position information

### DISCUSSION THREAD

**Discussion by Merovius** [yuin/goldmark#185] [Ideas]
Hi,

I was disappointed to see that the parser drops most of the source position information - it only seems to be preserving positional information for text nodes. I would have expected e.g. `ast.Link` to have a `text.Segment` field with `Start` pointing at the `[` and `Stop` pointing at the `)`. These would be nested, so the `Segment` of a child node would always `Start` and `Stop` inside the `Segment` (just as, for example, the `Link` has a child `Text` node, which already points at the text in-between the `[]`). i.e. all of this would work very similar to how `go/ast.Node.{Pos,End}` behaves.

Note, in particular, that this should make it trivial to implement #119, preserving the original markdown source. And personally, I want to be able to refer back and forth between HTML elements and the markdown that created them - I thought I would just need a custom `Renderer` that adds the relevant information to `data` attributes (which is already significant work). But without source positions, even that is futile.

In short: I think it would be great, if this information was preserved and reflected in `ast.Node` :)

**Comment by mickael-menu** [comment]
What a timely discussion, I encountered the same problem today!

For an additional use case, I wanted to locate a particular section of a Markdown document before extracting it from the original source (no AST modification). But I couldn't find a reliable way to get the node indices (`Lines` is not always filled) so I reverted back to regex parsing.

Apart from this, I really enjoyed working with this well-crafted library 👍

**Comment by frioux** [comment]
@yuin would you be interested in a PR that attempts to implement this functionality?  I'm trying to implement a basic markdown modification interface and this would make it much more feasible.

**Reply by mickael-menu** [reply]
I can say I'm still very interested in this feature 👍 I needed it again recently to locate links in a Markdown document and reverted to regexes for this.
